### DISCUSSION TITLE
Need help to integrate JS files with Breeze

### DISCUSSION THREAD

**Discussion by nkarthickannan** [breezefront/community#316] [1. Help]
Hello everyone, I'm new to the Breeze theme. I'm trying to learn how to integrate my custom module with Breeze blank theme. Basically, my module has 2 JS files. Hence, I created the breeze_default.xml file and added the components. Now I can see that my JS files are available in the Network tab of the browser.

However, a KO variable defined in my JS file is not working. Browser console throws this error:
![Screenshot 2024-08-15 at 10 28 27 AM](https://github.com/user-attachments/assets/f94d736d-2183-46d0-aab4-b42b629ba3f8)

Can anyone please help me to resolve this error ?

**Comment by sandipansaha** [comment]
Hi @nkarthickannan 

Could you please share the `breeze_default.xml` and JS files codes here, so we can try to help?

You can check the following docs as well whether you have missed any thing -

https://breezefront.com/docs/components
https://breezefront.com/docs/ui-components

Thanks

**Reply by nkarthickannan** [reply]
Hi @sandipansaha 

I read the documentation already. But I'm sure I'm missing something which I can't find on my own. So, I thought someone who is strong in Breeze can point out where I missed it.

I'm attaching the full module here:
[Archive.zip](https://github.com/user-attachments/files/16632355/Archive.zip)

**Comment by vovayatsyuk** [comment]
Maybe it's a conflict between widget and ui component names. UI component name is `ajaxwishlist`:

```js
Component.extend({
    component: 'ajaxwishlist',
    ...
})
```

widget name is `mage.ajaxWishlist`:

```js
$.widget('mage.ajaxWishlist', {
...
})
```

 That could be a issue in breeze. Try to rename your widget.

**Reply by nkarthickannan** [reply]
@vovayatsyuk 

Sure, I'll try changing it to match with each other. Just for understanding - What should be the correct value for the `component` variable ? If it needs to be matching with some other values, can you please highlight them ?

For example, I have references to this widget in multiple locations - `requirejs-config.js` file, `.phtml` file to include this, in this `$.widget` call and finally in Breeze's `component` variable. Do we need to match these values in any particular way to make sure it works ? On the other hand, in Breeze documentation, this `component` variable contains the actual path to the JS file in some places whereas some other sections contain the name of the component.

Your inputs in this regard will be really helpful to understand what happens behind the scenes.

Thanks

**Reply by vovayatsyuk** [reply]
`component` property must match the name you use in x-magento-init or data-mage-init or require declarations:

```
// component: 'Vendor_Module/js/component'
<div data-mage-init='{"Vendor_Module/js/component": {}}'></div>

// component: 'Vendor_Module/js/component'
<script type="text/x-magento-init">
{
    ".class-name": {
        "Vendor_Module/js/component": {}
    }
}
</script>

// component: 'Vendor_Module/js/component'
<script>
require(['Vendor_Module/js/component'], function () {
    //
});
</script>
```

**Reply by nkarthickannan** [reply]
@vovayatsyuk 

I made it to work finally !! However, I can't correlate how it works and what should be the exact reason. I can understand that the `component` name should match the `text/x-magento-init` call. However, considering the below screenshots, it is confusing. I hope it should make some sense for you. Your feedback is much appreciated !!

Working scenario:
![Screenshot 2024-08-16 at 9 59 13 PM](https://github.com/user-attachments/assets/2b793d51-af9d-478a-9e41-5435b9802e40)

Not working scenario - `component` name in JS file is matching with the inner `component` name in .phtml file
![Screenshot 2024-08-16 at 10 02 44 PM](https://github.com/user-attachments/assets/5c060a33-09c8-4b0c-b632-a498e54876df)

Not working scenario - `component` name in JS file is matching with the outer `components` name in .phtml file
![Screenshot 2024-08-16 at 10 06 09 PM](https://github.com/user-attachments/assets/631a1200-1f36-49a3-a5b7-ef507bac5e10)

**Reply by sandipansaha** [reply]
Hi @nkarthickannan 
It seems we are a bit confused with the different terminologies of a component.
Let me try to explain with a different example that I am recently working with.

My `phtml` file -

```html
<div id="event-booking-component" class="event-booking-wrapper">
    <div data-block="booking_component" data-bind="scope: 'event_booking'" style="visibility: hidden;">
        <h3 class="event-block-title">
            <?= $escaper->escapeHtml(__('Select your date')) ?>
        </h3>
        .......
        .......
    </div>
</div>
<script type="text/x-magento-init">
    {
        "[data-block='booking_component']": {
            "Magento_Ui/js/core/app": {
                "components": {
                    "event_booking": {
                        "component": "WilliamsCommerce_EventTicketManager/js/view/event/eventBooking",
                        "element": "[data-block='booking_component']",
                        "configuration": "<?= $escaper->escapeJs($block->getConfigurationJSON()) ?>"
                    }
                }
            }
        }
    }
</script>
``` 
![Screenshot from 2024-08-16 23-24-15](https://github.com/user-attachments/assets/40e94ce5-4fb8-4e76-b4a7-95bf60ccae03)
![Screenshot from 2024-08-16 23-26-45](https://github.com/user-attachments/assets/5d5eaf0b-b809-4d3b-8a7e-51a8a7cd5c3b)

My `breeze_default.xml` file -
```php
        <referenceBlock name="breeze.js">
            <arguments>
                <argument name="bundles" xsi:type="array">
                    <item name="events" xsi:type="array">
                        <item name="items" xsi:type="array">
                            <item name="WilliamsCommerce_EventTicketManager/js/view/event/eventBooking"
                                  xsi:type="array">
                                <item name="path" xsi:type="string">WilliamsCommerce_EventTicketManager/js/view/event/eventBooking</item>
                                <item name="enabled" xsi:type="helper" helper="Swissup\Breeze\Helper\Config::isEnabled">
                                    <param name="path">event_manager/general/enable</param>
                                </item>
                                <item name="import" xsi:type="array">
                                    <item name="WilliamsCommerce_EventTicketManager/js/view/event/calendarRenderer"
                                          xsi:type="string">WilliamsCommerce_EventTicketManager/js/view/event/calendarRenderer</item>
                                </item>
                                <item name="load" xsi:type="array">
                                    <item name="onReveal" xsi:type="array">
                                        <item name="0" xsi:type="string">#event-booking-component</item>
                                    </item>
                                </item>
                            </item>
                        </item>
                    </item>
                </argument>
            </arguments>
        </referenceBlock>
``` 
![Screenshot from 2024-08-16 23-36-39](https://github.com/user-attachments/assets/6f9807be-52be-46fe-9204-4fa8b084b2d0)

My `js` file
![Screenshot from 2024-08-16 23-39-45](https://github.com/user-attachments/assets/3f2cde02-2dd1-4143-aec9-f91cc7bbb185)

I hope it might help a bit on understanding those terms and how they are linked to each other and working.
@vovayatsyuk can correct me if I am wrong anywhere. :)

Thanks

**Comment by nkarthickannan** [comment]
@sandipansaha 

It's a great explanation and I really feel confident after seeing your screenshots. However, I found something different while experimenting with your explanations.

The `<item name=""` in `breeze_default.xml` does not seem to have any meaning in the code. It works with random characters as well without the need for matching with anything else. Please check the screen shot below. I placed `abcd` and `efgh` in the `<item name=""` sections; flushed the Magento caches and executed the `rm -rf pub/static/* var/cache/* var/page_cache/* var/tmp/* var/view_preprocessed/*` command to make sure that there are no impact because of the old code. It works without any issues. The main thing to note is that I didn't mention these `abcd` and `efgh` in any other places.
![Screenshot 2024-08-17 at 11 01 30 AM](https://github.com/user-attachments/assets/2215a393-e0cc-435e-80eb-a71dc11cace5)

In my opinion, `breeze_default.xml` file is only useful to make sure that our JS files are downloaded by the browser and it does not seem to have any relationship with the other JS files (I maybe completely wrong). Magento usually uses the XML files for building an array while processing in PHP. Similarly, this breeze_default.xml should be used to build an array of JS files that needs to be downloaded by the browser. I have not analysed the Breeze core module yet; but this is how it should have been implemented. @vovayatsyuk Your inputs will be highly valuable in this case.

So, the `component` variable which is mentioned in the Breeze documentation (screenshot below) must have some other logic running behind the scenes. It works when its value is matching with some XXXX and in other cases, it is not working. We need to find what that XXXX is. I'm just trying to explore on how it is coded so that we can get a better understanding.
![Screenshot 2024-08-17 at 11 16 18 AM](https://github.com/user-attachments/assets/3bbb2571-3e2b-4409-adb2-d1ceb4bef0ee)

@vovayatsyuk If you are free, please shed some light on this. It will be really helpful for understanding the Breeze module.

**Reply by vovayatsyuk** [reply]
You are correct! `breeze_default.xml` is used on the server side to build js bundles and provide information about components per each file or bundle.

Your components work fine when using random names in `breeze_default.xml` because of the `autoload` property. This property indicates that the file should be downloaded when the parent bundle is loaded. Since your components use a `default` bundle (bundle for common components) - breeze will download them on every page just because some other component triggered downloading of default.js.

It's worth noting that when bundling is enabled, your files will be merged into `default.js` and will work with random names in XML file even without the `autoload` property.

**Reply by nkarthickannan** [reply]
@vovayatsyuk Thank you for your explanation. Is it wise to use the `autoload` property in my code or not ?

**Reply by vovayatsyuk** [reply]
I think you don't need it. We use `autload` to include js file that is not linked to any component. The example [theme.js files](https://github.com/breezefront/theme-frontend-breeze-blank/blob/master/Magento_Theme/layout/default_head_blocks.xml#L23-L34)
